### ISSUE TITLE
Integrate nvidia-DALI?

### ISSUE THREAD

**Issue by ashok-arjun** [ContinualAI/avalanche#892]
[NVIDIA Dali](https://github.com/NVIDIA/DALI) is a GPU-accelerated library that speeds up data-processing pipelines by as high as 67% for large datasets like imagenet (some benchmarks [here](https://developer.nvidia.com/blog/case-study-resnet50-dali/) and [here](https://github.com/tanglang96/DataLoaders_DALI) (the speedup is higher with a larger number of workers).

Integrating this library into avalanche would make it appealing for users. I'd be happy to take this up.

I'd also like to point out that this is further useful in the class-incremental learning setup where a major overhead is the additional set of data at every experience. It would offer large speedups for huge datasets like imagenet / core-50 where data processing takes up a lot of time

**Comment by AntonioCarta** [comment]
Thanks, this is an important issue. I will add [ffcv](https://ffcv.io/) as another option to evaluate.

There are a couple of isses to solve to guarantee fast dataloading:
- ensure default dataloaders created by the library are fast (yet they should be general).
- figure out how to support data loading in online learning. Since each experience has only one sample at a time, we create the dataloader from scratch every time. This means having `num_workers>0` is useless. It's the hard part of this problem, and also the setting where we have the largest slowdown.

> Some recent libraries have also integrated it and found it to be useful (see this section of a recent library).

Wow, I knew pytorch transforms were not optimized but I did not expect such a large speedup.

**Comment by ashok-arjun** [comment]
Yes, I agree with the online CL part. Recent works use batch_size as 10 for online CL - but I think we should make it generalized to be always 1?

With `ffcv`, I've been trying it out recently, and there are so many environment-related problems. `DALI` is more stable.

**Comment by AntonioCarta** [comment]
can you try to test both with Avalanche by overriding `make_train_dataloader` and `make_eval_dataloader` and let us know if anything breaks? This small experiment will be useful to decide which one to use.

> With ffcv, I've been trying it out recently, and there are so many environment-related problems. DALI is more stable.

This is an important point. Are you on linux/windows/mac?

Ideally, the best solution would be transparent to the user (dropin replacement for pytorch dataloaders) and easy to package, otherwise we will break the user's environments very easily.

**Comment by ashok-arjun** [comment]
Well actually the current integrations need a lot of new code to be completely the same as the original loaders - an example is transforms. For every torchvision transform, there is an equivalent transform in the Dali library, as well as in the FFCV library (understandable, since they accelerate the transforms).

As for `ffcv`, I'm on linux right now and facing a lot of issues. Plus, it is only available for Python3.9 + CUDA11.3 so I had to create a new environment to try it (inspite of which it didn't work). I feel its best to wait until a stable release appears.

As for `dali`, the releases are stable and it is actively in development. It would be useful for large-scale experiments.

So I'll do some controlled experiments with `dali` and come back with the increase observed in my settings (CPU cores and number of workers). I'll restrict myself to using a single GPU as only that's supported by Avalanche.

**Comment by lrzpellegrini** [comment]
FFCV support has now been added in the master branch (it is NOT included in the latest pip release, v0.4.0). See the examples for more details.

**Comment by AntonioCarta** [comment]
I think we can close this now and open more specific issues if needed to track subsequent work.
